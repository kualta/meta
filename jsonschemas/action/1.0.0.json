{
  "type": "object",
  "properties": {
    "$schema": {
      "type": "string",
      "const": "https://json-schemas.lens.dev/actions/1.0.0.json"
    },
    "lens": {
      "type": "object",
      "properties": {
        "id": {
          "$ref": "#/$defs/NonEmptyString",
          "description": "A unique identifier that in storages like IPFS ensures the uniqueness of the metadata URI. Use a UUID if unsure."
        },
        "name": {
          "$ref": "#/$defs/NonEmptyString",
          "description": "The name of the Action."
        },
        "title": {
          "$ref": "#/$defs/NonEmptyString",
          "description": "The human-friendly title for the Action."
        },
        "description": {
          "$ref": "#/$defs/NonEmptyString",
          "description": "Markdown formatted description of the Action."
        },
        "authors": {
          "type": "array",
          "items": {
            "type": "string",
            "format": "email"
          },
          "minItems": 1,
          "description": "List of authors email addresses."
        },
        "source": {
          "type": "string",
          "minLength": 6,
          "format": "uri",
          "description": "The link to the Action source code. Typically a GitHub repository."
        },
        "configureParams": {
          "type": "array",
          "items": {
            "type": "object",
            "properties": {
              "key": {
                "type": "string",
                "minLength": 66,
                "maxLength": 66,
                "pattern": "^0x[0-9a-fA-F]{64}$",
                "description": "A unique 32 bytes long hexadecimal string key."
              },
              "name": {
                "$ref": "#/$defs/NonEmptyString",
                "description": "The human-readable name of the parameter."
              },
              "type": {
                "type": "string",
                "description": "The human-readable ABI description of the parameter."
              }
            },
            "required": [
              "key",
              "name",
              "type"
            ],
            "additionalProperties": false
          },
          "description": "An optional list of `ContractKeyValuePairDescriptor` that describe the `params` argument of the `configure` function."
        },
        "executeParams": {
          "type": "array",
          "items": {
            "$ref": "#/properties/lens/properties/configureParams/items"
          },
          "description": "A list of `ContractKeyValuePairDescriptor` that describe the `params` argument of the `execute` function."
        },
        "setDisabledParams": {
          "type": "array",
          "items": {
            "$ref": "#/properties/lens/properties/configureParams/items"
          },
          "description": "An optional list of `ContractKeyValuePairDescriptor` that describe the `params` argument of the `setDisabledParams` function."
        }
      },
      "required": [
        "id",
        "name",
        "title",
        "description",
        "authors",
        "source",
        "executeParams"
      ],
      "additionalProperties": false
    },
    "signature": {
      "type": "string",
      "minLength": 1,
      "description": "A cryptographic signature of the Lens metadata."
    }
  },
  "required": [
    "$schema",
    "lens"
  ],
  "additionalProperties": false,
  "$defs": {
    "NonEmptyString": {
      "type": "string",
      "minLength": 1
    },
    "MetadataAttribute": {
      "anyOf": [
        {
          "type": "object",
          "properties": {
            "type": {
              "type": "string",
              "const": "Boolean"
            },
            "key": {
              "$ref": "#/$defs/NonEmptyString",
              "description": "The attribute's unique identifier."
            },
            "value": {
              "type": "string",
              "enum": [
                "true",
                "false"
              ],
              "description": "A JS boolean value serialized as string. It's consumer responsibility to parse it."
            }
          },
          "required": [
            "type",
            "key",
            "value"
          ],
          "additionalProperties": false
        },
        {
          "type": "object",
          "properties": {
            "type": {
              "type": "string",
              "const": "Date"
            },
            "key": {
              "$ref": "#/$defs/NonEmptyString",
              "description": "The attribute's unique identifier."
            },
            "value": {
              "type": "string",
              "format": "date-time",
              "description": "A valid ISO 8601 date string.  It's consumer responsibility to parse it."
            }
          },
          "required": [
            "type",
            "key",
            "value"
          ],
          "additionalProperties": false
        },
        {
          "type": "object",
          "properties": {
            "type": {
              "type": "string",
              "const": "Number"
            },
            "key": {
              "$ref": "#/$defs/NonEmptyString",
              "description": "The attribute's unique identifier."
            },
            "value": {
              "$ref": "#/$defs/NonEmptyString",
              "description": "A valid JS number serialized as string. It's consumer responsibility to parse it."
            }
          },
          "required": [
            "type",
            "key",
            "value"
          ],
          "additionalProperties": false
        },
        {
          "type": "object",
          "properties": {
            "type": {
              "type": "string",
              "const": "String"
            },
            "key": {
              "$ref": "#/$defs/NonEmptyString",
              "description": "The attribute's unique identifier."
            },
            "value": {
              "$ref": "#/$defs/NonEmptyString",
              "description": "Any string value."
            }
          },
          "required": [
            "type",
            "key",
            "value"
          ],
          "additionalProperties": false
        },
        {
          "type": "object",
          "properties": {
            "type": {
              "type": "string",
              "const": "JSON"
            },
            "key": {
              "$ref": "#/$defs/NonEmptyString",
              "description": "The attribute's unique identifier."
            },
            "value": {
              "$ref": "#/$defs/NonEmptyString",
              "description": "A JSON string. It's consumer responsibility to validate and parse it."
            }
          },
          "required": [
            "type",
            "key",
            "value"
          ],
          "additionalProperties": false
        }
      ]
    }
  },
  "$schema": "http://json-schema.org/draft-07/schema#"
}
