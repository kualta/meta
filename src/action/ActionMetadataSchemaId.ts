import { SchemasRoot } from '../constants';

/**
 * The Lens Action metadata schema id.
 */
export enum ActionMetadataSchemaId {
  LATEST = `${SchemasRoot}/actions/1.0.0.json`,
}
