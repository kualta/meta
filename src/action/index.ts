export * from './ActionMetadataSchemaId';
export * from './ActionMetadataSchema';
