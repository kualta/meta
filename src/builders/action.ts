import { v4 } from 'uuid';

import {
  type ActionMetadata,
  type ActionMetadataDetails,
  ActionMetadataSchema,
  ActionMetadataSchemaId,
} from '../action/index.js';
import type { Prettify } from '../utils.js';
import { evaluate } from './ValidationError';
import type { RecursiveUnbrand } from './utils';

/**
 * @private
 * @privateRemarks MUST stay very @private to produce usable docs
 */
type ActionDetails = Prettify<RecursiveUnbrand<Omit<ActionMetadataDetails, 'id'>>>;

export type ActionOptions = ActionDetails & {
  /**
   * A unique identifier that in storages like IPFS ensures the uniqueness of the metadata URI.
   *
   * @defaultValue a UUID
   */
  id?: string;
};

/**
 * Creates a valid ActionMetadata.
 *
 * @category Compose
 * @param input - Use your IDE suggestions for an enhanced development experience
 *
 * @example
 * ```ts
 * const metadata = action({
 *   name: 'MyPostAction',
 *   description: 'Get ready for the future of social interaction!',
 *   authors: ['awesome-dev@lens.xyz'],
 *   source: 'https://github.com/foobarbaz/my-open-action',
 *   executeParams: [
 *     {
 *       key: '83b03721c31435de384ca9cc3799cdfa10e19123e6a48dbf11589e36a0ef0b95',
 *       name: 'recipient',
 *       type: 'address',
 *     },
 *     {
 *       key: 'e4ddce99520beb836fc879fc2a7d0743f784ffb925f96544bb5ff27c441c8c99',
 *       name: 'amount',
 *       type: 'address token, uint256 value',
 *     }
 *   ]
 * });
 * ```
 */
export function action({ id = v4(), ...others }: ActionOptions): ActionMetadata {
  return evaluate(
    ActionMetadataSchema.safeParse({
      $schema: ActionMetadataSchemaId.LATEST,
      lens: {
        id,
        ...others,
      },
    }),
  );
}
