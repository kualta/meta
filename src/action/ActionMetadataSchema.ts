import { z } from 'zod';

import {
  type ContractKeyValuePairDescriptor,
  ContractKeyValuePairDescriptorSchema,
  type Markdown,
  MarkdownSchema,
  NonEmptyStringSchema,
  type Signature,
  SignatureSchema,
  type URI,
  URISchema,
} from '../primitives';
import { ActionMetadataSchemaId } from './ActionMetadataSchemaId';

export type ActionMetadataDetails = {
  /**
   * A unique identifier that in storages like IPFS ensures the uniqueness of the metadata URI. Use a UUID if unsure.
   */
  id: string;
  /**
   * The name of the Action.
   */
  name: string;
  /**
   * The human-friendly title for the Action.
   */
  title: string;
  /**
   * Markdown formatted description of the Action.
   *
   * It should explain what this Action does, how to use it, examples just
   * like you would if you were building an NPM package.
   * The more detail the better to allow dApp to easily integrate it.
   */
  description: Markdown;
  /**
   * List of authors email addresses.
   */
  authors: string[];
  /**
   * The link to the Action source code. Typically a GitHub repository.
   */
  source: URI;
  /**
   * An optional list of {@link ContractKeyValuePairDescriptor} that describe the `params` argument of the `configure` function.
   *
   * ```sol
   * function configure(address originalMsgSender, address feed, uint256 postId, KeyValue[] calldata params)
   *   external
   *   returns (bytes memory);
   * ```
   *
   * This will be used to know how to encode and decode the params for the `configuration` function.
   *
   * @see https://abitype.dev/api/human
   */
  configureParams?: ContractKeyValuePairDescriptor[];
  /**
   * A list of {@link ContractKeyValuePairDescriptor} that describe the `params` argument of the `execute` function.
   *
   * ```sol
   * function execute(address originalMsgSender, address feed, uint256 postId, KeyValue[] calldata params)
   *   external
   *   returns (bytes memory);
   * ```
   *
   * This will be used to know how to encode the calldata for the `execute` function.
   */
  executeParams?: ContractKeyValuePairDescriptor[];
  /**
   * An optional list of {@link ContractKeyValuePairDescriptor} that describe the `params` argument of the `setDisabledParams` function.
   *
   * ```sol
   * function setDisabled(
   *   address originalMsgSender,
   *   address feed,
   *   uint256 postId,
   *   bool isDisabled,
   *   KeyValue[] calldata params
   * ) external returns (bytes memory);
   * ```
   *
   * This will be used to know how to encode the calldata for the `setDisabled` function.
   */
  setDisabledParams?: ContractKeyValuePairDescriptor[];
};

const ActionMetadataDetailsSchema: z.ZodType<ActionMetadataDetails, z.ZodTypeDef, object> =
  z.object({
    id: NonEmptyStringSchema.describe(
      'A unique identifier that in storages like IPFS ensures the uniqueness of the metadata URI. Use a UUID if unsure.',
    ),
    name: NonEmptyStringSchema.describe('The name of the Action.'),
    title: NonEmptyStringSchema.describe('The human-friendly title for the Action.'),
    description: MarkdownSchema.describe('Markdown formatted description of the Action.'),
    authors: z
      .array(z.string().email('Authors list should only contain valid emails'))
      .min(1, 'You must supply at least one author')
      .describe('List of authors email addresses.'),
    source: URISchema.describe(
      'The link to the Action source code. Typically a GitHub repository.',
    ),
    configureParams: z
      .array(ContractKeyValuePairDescriptorSchema)
      .optional()
      .describe(
        'An optional list of `ContractKeyValuePairDescriptor` that describe the `params` argument of the `configure` function.',
      ),
    executeParams: z
      .array(ContractKeyValuePairDescriptorSchema)
      .describe(
        'A list of `ContractKeyValuePairDescriptor` that describe the `params` argument of the `execute` function.',
      ),
    setDisabledParams: z
      .array(ContractKeyValuePairDescriptorSchema)
      .optional()
      .describe(
        'An optional list of `ContractKeyValuePairDescriptor` that describe the `params` argument of the `setDisabledParams` function.',
      ),
  });

export type ActionMetadata = {
  $schema: ActionMetadataSchemaId.LATEST;
  lens: ActionMetadataDetails;
  signature?: Signature;
};

export const ActionMetadataSchema: z.ZodType<ActionMetadata, z.ZodTypeDef, object> = z.object({
  $schema: z.literal(ActionMetadataSchemaId.LATEST),
  lens: ActionMetadataDetailsSchema,
  signature: SignatureSchema.optional(),
});
