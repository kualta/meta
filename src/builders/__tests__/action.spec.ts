import { describe, expect, it } from 'vitest';

import { action } from '../action.js';

describe(`Given the "${action.name}" metadata builder`, () => {
  describe('when using it', () => {
    it('should return a valid AccountMetadata', () => {
      const metadata = action({
        name: 'My Open Action',
        title: 'This is my Open Action',
        description: 'Get ready for the future of social interaction!',
        authors: ['awesome-dev@lens.xyz'],
        source: 'https://github.com/foobarbaz/my-open-action',
        configureParams: [
          {
            key: '0xc8a06abcb0f2366f32dc2741bdf075c3215e3108918311ec0ac742f1ffd37f49',
            name: 'account',
            type: 'address',
          },
          {
            key: '0xc8a06abcb0f2366f32dc2741bdf075c3215e3108918311ec0ac742f1ffd37f49',
            name: 'recipients',
            type: 'address[]',
          },
        ],
        executeParams: [
          {
            key: '0xc8a06abcb0f2366f32dc2741bdf075c3215e3108918311ec0ac742f1ffd37f49',
            name: 'amount',
            type: 'uint256',
          },
          {
            key: '0xc8a06abcb0f2366f32dc2741bdf075c3215e3108918311ec0ac742f1ffd37f49',
            name: 'recipients',
            type: '(address, uint256)[]',
          },
        ],
        setDisabledParams: [
          {
            key: '0xc8a06abcb0f2366f32dc2741bdf075c3215e3108918311ec0ac742f1ffd37f49',
            name: 'amount',
            type: 'uint256',
          },
          {
            key: '0xc8a06abcb0f2366f32dc2741bdf075c3215e3108918311ec0ac742f1ffd37f49',
            name: 'recipients',
            type: '(address, uint256)[]',
          },
        ],
      });

      expect(metadata).toMatchSnapshot({
        lens: {
          id: expect.any(String),
        },
      });
    });
  });
});
