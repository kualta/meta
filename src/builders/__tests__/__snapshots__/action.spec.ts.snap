// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Given the "action" metadata builder > when using it > should return a valid AccountMetadata 1`] = `
{
  "$schema": "https://json-schemas.lens.dev/actions/1.0.0.json",
  "lens": {
    "authors": [
      "awesome-dev@lens.xyz",
    ],
    "configureParams": [
      {
        "key": "0xc8a06abcb0f2366f32dc2741bdf075c3215e3108918311ec0ac742f1ffd37f49",
        "name": "account",
        "type": "address",
      },
      {
        "key": "0xc8a06abcb0f2366f32dc2741bdf075c3215e3108918311ec0ac742f1ffd37f49",
        "name": "recipients",
        "type": "address[]",
      },
    ],
    "description": "Get ready for the future of social interaction!",
    "executeParams": [
      {
        "key": "0xc8a06abcb0f2366f32dc2741bdf075c3215e3108918311ec0ac742f1ffd37f49",
        "name": "amount",
        "type": "uint256",
      },
      {
        "key": "0xc8a06abcb0f2366f32dc2741bdf075c3215e3108918311ec0ac742f1ffd37f49",
        "name": "recipients",
        "type": "(address, uint256)[]",
      },
    ],
    "id": Any<String>,
    "name": "My Open Action",
    "setDisabledParams": [
      {
        "key": "0xc8a06abcb0f2366f32dc2741bdf075c3215e3108918311ec0ac742f1ffd37f49",
        "name": "amount",
        "type": "uint256",
      },
      {
        "key": "0xc8a06abcb0f2366f32dc2741bdf075c3215e3108918311ec0ac742f1ffd37f49",
        "name": "recipients",
        "type": "(address, uint256)[]",
      },
    ],
    "source": "https://github.com/foobarbaz/my-open-action",
    "title": "This is my Open Action",
  },
}
`;
